test/utilities.py

```diff
@@ -1,8 +1,142 @@
+import hashlib
 import inspect
-import sys
 import os
-import hashlib
-import datetime
+import pathlib
+import shutil
+import sys
+import tempfile
+import warnings
+from datetime import datetime
+from subprocess import check_call, check_output
+
+from fpdf.template import Template
+
+
+QPDF_AVAILABLE = bool(shutil.which("qpdf"))
+if not QPDF_AVAILABLE:
+    warnings.warn(
+        "qpdf command not available on the $PATH, falling-back to hash-based comparisons in tests"
+    )
+
+
+class TempFile:
+    """docstring for TempFile"""
+
+    def __init__(self, name, handle):
+        super(TempFile, self).__init__()
+        self.name = name
+        self.handle = handle
+
+    def read_whole_file(self):
+        ret = []
+        while True:
+            rval = self.handle.read(10_000)
+            if not rval or rval == b"":
+                break
+            ret.append(rval)
+        return b"".join(ret)
+
+    def rwf(self):
+        return self.read_whole_file()
+
+    def close(self):
+        self.handle.close()
+
+
+class TempFileCtx:
+    """docstring for TempFileCtxs"""
+
+    def __init__(self, prefix, delete, suffix):
+        super(TempFileCtx, self).__init__()
+        self.prefix = prefix
+        self.delete = delete
+        self.suffix = suffix
+
+    def random(self):
+        return os.urandom(24).hex()
+
+    def get_os_tmp(self):
+        return tempfile.gettempdir()
+
+    def get_tmp_file_name_base(self):
+        return self.prefix + self.random() + self.suffix
+
+    def get_tmp_file_name(self):
+        name = self.get_tmp_file_name_base()
+        return os.path.join(self.get_os_tmp(), name)
+
+    def __enter__(self):
+        self.abs_name = self.get_tmp_file_name()
+        pathlib.Path(self.abs_name).touch()
+        tmp_file = TempFile(self.abs_name, open(self.abs_name, "rb"))
```

Comment:
This is file is never closed!

Suggestion:
```python
 with open(self.abs_name, "rb") as tmp_file
        self.tmp_file = TempFile(self.abs_name, tmp_file)
        return self.tmp_file
```

Reply:
should be closed in exit method